File: frontend/src/api/tasks.js
import axios from "axios";

const BASE_URL = import.meta.env.VITE_API_URL || "https://getitdone-app.onrender.com";

// -------------------------
// GET TASKS (JWT REQUIRED)
// -------------------------
export const getTasks = async () => {
  const token = localStorage.getItem("token");

  const res = await axios.get(`${BASE_URL}/tasks`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  return res.data;
};

// -------------------------
// CREATE TASK (JWT REQUIRED)
// -------------------------
export const createTask = async (title, description, assigned_to_id, due_date) => {
  const token = localStorage.getItem("token");

  const res = await axios.post(
    `${BASE_URL}/tasks`,
    {
      title,
      description,
      assigned_to_id,
      due_date
    },
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return res.data;
};

// -------------------------
// UPDATE TASK (JWT REQUIRED)
// -------------------------
export const updateTask = async (taskId, updates) => {
  const token = localStorage.getItem("token");

  const res = await axios.put(
    `${BASE_URL}/tasks/${taskId}`,
    updates,
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return res.data;
};

// -------------------------
// DELETE TASK (JWT REQUIRED)
// -------------------------
export const deleteTask = async (taskId) => {
  const token = localStorage.getItem("token");

  const res = await axios.delete(`${BASE_URL}/tasks/${taskId}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  return res.data;
};

// -------------------------
// TIP TASK (JWT REQUIRED)
// -------------------------
export const tipTask = async (taskId, amount) => {
  const token = localStorage.getItem("token");

  const res = await axios.post(
    `${BASE_URL}/tasks/${taskId}/tip`,
    { amount },
    {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  return res.data;
};

// -------------------------
// GET TASK EVENTS (JWT REQUIRED)
// -------------------------
export const getTaskEvents = async (taskId) => {
  const token = localStorage.getItem("token");

  const res = await axios.get(`${BASE_URL}/tasks/${taskId}/events`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  return res.data;
};

// -------------------------
// SUBTASKS
// -------------------------
export const createSubtask = async (taskId, title) => {
  const token = localStorage.getItem("token");
  const res = await axios.post(
    `${BASE_URL}/tasks/${taskId}/subtasks`,
    { title },
    { headers: { Authorization: `Bearer ${token}` } }
  );
  return res.data;
};

export const updateSubtask = async (subtaskId, updates) => {
  const token = localStorage.getItem("token");
  const res = await axios.put(
    `${BASE_URL}/subtasks/${subtaskId}`,
    updates,
    { headers: { Authorization: `Bearer ${token}` } }
  );
  return res.data;
};

export const deleteSubtask = async (subtaskId) => {
  const token = localStorage.getItem("token");
  const res = await axios.delete(
    `${BASE_URL}/subtasks/${subtaskId}`,
    { headers: { Authorization: `Bearer ${token}` } }
  );
  return res.data;
};

File: backend/routers/tasks.py
from fastapi import APIRouter, Depends, BackgroundTasks
from sqlalchemy.orm import Session
from uuid import UUID

from database import get_db
from models import User
from schemas.tasks import TaskCreate, TaskUpdate, TaskTipRequest, SubtaskCreate, SubtaskUpdate
from dependencies import get_current_user
from services import tasks_service

router = APIRouter()

@router.post("/tasks")
def create_task(
    task: TaskCreate,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.create_task(task, background_tasks, db, current_user)

@router.get("/tasks")
def get_tasks(
    filter_by: str = 'all',
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.get_tasks(filter_by, db, current_user)

@router.get("/tasks/{task_id}/events")
def get_task_events(task_id: UUID, db: Session = Depends(get_db), current_user: User = Depends(get_current_user)):
    return tasks_service.get_task_events(task_id, db, current_user)

@router.put("/tasks/{task_id}")
def update_task(
    task_id: UUID,
    task_update: TaskUpdate,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.update_task(task_id, task_update, background_tasks, db, current_user)

@router.delete("/tasks/{task_id}")
def delete_task(
    task_id: UUID,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.delete_task(task_id, background_tasks, db, current_user)

@router.post("/tasks/{task_id}/tip")
def tip_task(
    task_id: UUID,
    tip_req: TaskTipRequest,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.tip_task(task_id, tip_req, background_tasks, db, current_user)

@router.post("/tasks/{task_id}/subtasks")
def create_subtask(
    task_id: UUID,
    subtask: SubtaskCreate,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.create_subtask(task_id, subtask, background_tasks, db, current_user)

@router.put("/tasks/{task_id}/subtasks/{subtask_id}")
def update_subtask(
    task_id: UUID,
    subtask_id: UUID,
    subtask_update: SubtaskUpdate,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.update_subtask(task_id, subtask_id, subtask_update, background_tasks, db, current_user)

@router.delete("/tasks/{task_id}/subtasks/{subtask_id}")
def delete_subtask(
    task_id: UUID,
    subtask_id: UUID,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    return tasks_service.delete_subtask(task_id, subtask_id, background_tasks, db, current_user)
